File: benchmarks/cross-language/scala/fork_join.scala
// Scala Akka Fork-Join Benchmark (Savina-style)
import akka.actor.{Actor, ActorRef, ActorSystem, Props}
import scala.concurrent.duration._
import scala.concurrent.{Await, Promise}

case class Work(value: Int)
case object Done
case class Result(processed: Int)

class WorkerActor(parent: ActorRef) extends Actor {
  var processed = 0

  def receive = {
    case Work(_) =>
      processed += 1
    case Done =>
      parent ! Result(processed)
  }
}

class CoordinatorActor(promise: Promise[Int], numWorkers: Int, messagesPerWorker: Int) extends Actor {
  var workers: Array[ActorRef] = null
  var resultsReceived = 0
  var totalProcessed = 0
  var startTime = 0L

  override def preStart(): Unit = {
    // Create workers
    workers = Array.tabulate(numWorkers)(i =>
      context.actorOf(Props(classOf[WorkerActor], self), s"worker$i")
    )

    startTime = System.nanoTime()

    // Send messages round-robin
    val total = numWorkers * messagesPerWorker
    for (i <- 0 until total) {
      workers(i % numWorkers) ! Work(i)
    }

    // Signal workers to report
    workers.foreach(_ ! Done)
  }

  def receive = {
    case Result(processed) =>
      totalProcessed += processed
      resultsReceived += 1
      if (resultsReceived == numWorkers) {
        promise.success(totalProcessed)
        context.system.terminate()
      }
  }
}

object ForkJoinBenchmark extends App {
  val numWorkers = 8
  val totalMessages = sys.env.get("BENCHMARK_MESSAGES").flatMap(s => scala.util.Try(s.toInt).toOption).getOrElse(100000)
  val messagesPerWorker = totalMessages / numWorkers
  val total = numWorkers * messagesPerWorker

  println("=== Scala Akka Fork-Join Throughput Benchmark ===")
  println(s"Workers: $numWorkers, Messages: $total\n")

  val system = ActorSystem("ForkJoin")
  val promise = Promise[Int]()

  val startTime = System.nanoTime()
  val coordinator = system.actorOf(Props(classOf[CoordinatorActor], promise, numWorkers, messagesPerWorker))

  val totalProcessed = Await.result(promise.future, 120.seconds)
  val endTime = System.nanoTime()

  val elapsed = (endTime - startTime).toDouble / 1e9

  if (totalProcessed != total) {
    println(s"VALIDATION FAILED: expected $total, got $totalProcessed")
  }

  val throughput = total / elapsed / 1e6
  val nsPerMsg = elapsed * 1e9 / total

  println(f"ns/msg:         $nsPerMsg%.2f")
  println(f"Throughput:     $throughput%.2f M msg/sec")

  Await.result(system.whenTerminated, 10.seconds)
}

File: benchmarks/cross-language/rust/src/fork_join.rs
// Rust Fork-Join Benchmark (Savina-style)
use std::sync::mpsc::{channel, Sender, Receiver};
use std::thread;
use std::time::Instant;

const NUM_WORKERS: usize = 8;

fn get_messages() -> usize {
    std::env::var("BENCHMARK_MESSAGES")
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(100_000)
}

fn worker(rx: Receiver<i32>) -> usize {
    let mut count = 0;
    while rx.recv().is_ok() {
        count += 1;
    }
    count
}

fn main() {
    let total = get_messages();
    println!("=== Rust Fork-Join Throughput Benchmark ===");
    println!("Workers: {}, Messages: {}\n", NUM_WORKERS, total);

    // Create workers
    let mut senders: Vec<Sender<i32>> = Vec::with_capacity(NUM_WORKERS);
    let mut handles = Vec::with_capacity(NUM_WORKERS);

    for _ in 0..NUM_WORKERS {
        let (tx, rx) = channel();
        senders.push(tx);
        handles.push(thread::spawn(move || worker(rx)));
    }

    let start = Instant::now();

    // Send messages round-robin
    for i in 0..total {
        senders[i % NUM_WORKERS].send(i as i32).unwrap();
    }

    // Signal workers to finish
    drop(senders);

    // Collect results
    let mut total_processed = 0;
    for h in handles {
        total_processed += h.join().unwrap();
    }

    let elapsed = start.elapsed().as_secs_f64();

    if total_processed != total {
        println!("VALIDATION FAILED: expected {}, got {}", total, total_processed);
    }

    let throughput = total as f64 / elapsed / 1e6;
    let ns_per_msg = elapsed * 1e9 / total as f64;

    println!("ns/msg:         {:.2}", ns_per_msg);
    println!("Throughput:     {:.2} M msg/sec", throughput);
}
